notebook/exploration-reports/2022.03-ipld-url-scheme.md

```diff
@@ -0,0 +1,106 @@
+IPLD URL Scheme v0
+==================
+
+This is a proposed IPLD URL scheme spec, focusing on a minimap viable product which doesn't address advanced use cases such as lenses or selectors.
+
+**Author**: Mauve Signeaver ([@RangerMauve])
+
+---
+
+tl;dr
+
+```
+GET ipld://bafy....etc/some/path => {some: data}
+PUT ipld://bafy....etc/some/path {some: data} => ipld://bafy...etc/
+```
+
+## Objective
+
+`ipfs://` exists out in the wild to provide ways to link to files, and some web browsers integrate IPFS URLs to load and author content.
+This spec shows how a similar URL scheme can exist for IPLD data.
+Part of the goal of the spec is to make it easy to reuse more or less the same APIs between software that handles `ipld://` URLs natively, HTTP gateways like the [IPFS gateway](https://docs.ipfs.io/concepts/ipfs-gateway/) and CLIs.
+With that in mind, there should be a clear mapping between HTTP-like APIs via gateways and URL handlers, and CLI arguments.
+In order to make it easier to come to an initial spec, we will focus on a minimal set of features and make subsequent specs with support for features like ADLs, Schemas, and advanced selectors.
+
+## Initial Functionality
+
+The initial functionality focuses on being able to GET and PUT IPLD data to and from a URL.
+For now, this will be done by encoding individual blocks of IPLD data using codecs with a default to the [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/) codec.
+Some initial support for pathing is also included for convenience.
+
+### GET `ipld://{CID}/{Path} Headers {Accept="application/vnd.ipld.dag-json"} => Data`
+
+The most basic mapping is to get some data from a CID.
+
+The `CID` parameter in the URL MUST be in the CIDv1 format or any other format that is not case sensitive.
+This is necessary since the URL scheme considers the `hostname` portion to be case-insensitive and many URL parsers (such as the one in Chrome) will automatically convert it to lowercase. 
+
+`Data` is the response body which contaisn the encoded IPLD data for the given CID+Path.
+Note that any `Link`s within the IPLD data will not be resolved, and you'll need to conver them to URLs and fetch additional data yourself, or add path segments to the URL to traverse into them.
+
+The `Accept` header may be optionally provided to specify the codec that should be used in the response.
+This follows the same conventions as in the [IPFS gateway](https://github.com/ipfs/go-ipfs/pull/8758).
+Other codecs that may be targetted are DAG-CBOR.
+TODO: `application/vnd.ipld.dag-json` or just `application/json`?
```

Comment:
My only worry with regards to using `application/vnd.ipld.dag-json` is that it wouldn't get recognized by common browser tools like syntax highlighting extensions and the such. I think it could still work with `await (await fetch(url)).json()` in browsers, though.

Reply:
My only worry with regards to using `application/vnd.ipld.dag-json` is that it wouldn't get recognized by common browser tools like syntax highlighting extensions and the such. I think it could still work with `await (await fetch(url)).json()` in browsers, though.

Reply:
@RangerMauve fair point! so.. `application/json` everywhere + also support `application/vnd.ipld.dag-json` in requests? that should maximize reuse of things that have nice UX/support for JSON.

Reply:
Yeah, I'm into that. I'll add it to the doc. 😁